File: examples/csharp/TimeMachine.cs
using System.Text;
using System.Text.Json;
using WickraTerminal;

/// <summary>
/// The time-machine scenario: rewind a recorded feed and watch state re-fold.
///
/// <c>Seek</c> throws the folded state away and rebuilds it from the recording,
/// so a rewind is deterministic rather than approximate — which is what makes a
/// recording more than a slow synthetic feed. Nothing here is C#-specific: it is
/// four JSON commands, and every binding drives the same four.
/// </summary>
internal static class TimeMachine
{
    private const int Trades = 6;

    /// <summary>The recorded feed, as the JSON array a Replay source takes.</summary>
    private static string Config()
    {
        var events = new StringBuilder("[");
        for (int i = 0; i < Trades; i++)
        {
            if (i > 0)
            {
                events.Append(',');
            }
            events.Append("{\"type\":\"trade\",\"symbol\":{\"base\":\"BTC\",\"quote\":\"USDT\"},")
                  .Append("\"price\":\"").Append(100 + i).Append("\",\"quantity\":\"1\",")
                  .Append("\"aggressor\":\"Buy\",\"timestamp\":").Append(i + 1).Append('}');
        }
        events.Append(']');
        string feed = JsonSerializer.Serialize(events.ToString());
        return "{\"sources\":[{\"Replay\":{\"dataset\":" + feed + "}}]," +
               "\"layout\":{\"panels\":[{\"kind\":\"Chart\",\"rect\":{\"x\":0,\"y\":0,\"w\":100,\"h\":100}}]}}";
    }

    /// <summary>The chart panel's last price, out of a frame.</summary>
    private static double LastPrice(string frame)
    {
        using var parsed = JsonDocument.Parse(frame);
        foreach (var panel in parsed.RootElement.GetProperty("panels").EnumerateArray())
        {
            if (panel.GetProperty("panel").GetString() == "chart")
            {
                return panel.GetProperty("last").GetDouble();
            }
        }
        return 0;
    }

    public static void Run()
    {
        using var term = new Terminal(Config());
        term.Command("{\"type\":\"Subscribe\",\"source\":0,\"symbol\":\"BTC/USDT\"}");

        string raw = string.Empty;
        for (int i = 0; i < Trades; i++)
        {
            raw = term.Command("{\"type\":\"Tick\"}");
        }
        Console.WriteLine($"played to the end:   last = {LastPrice(raw)}");

        Console.WriteLine("position:            " +
            term.Command("{\"type\":\"ReplayPosition\",\"source\":0}"));

        // Rewind to just after the second trade. The state is rebuilt from the
        // recording rather than restored from a snapshot, which is why a rewind
        // lands on exactly the frame the forward pass had at that point.
        raw = term.Command("{\"type\":\"Seek\",\"source\":0,\"index\":2}");
        Console.WriteLine($"rewound to index 2:  last = {LastPrice(raw)}");

        // And forward again from there, over the same events.
        raw = term.Command("{\"type\":\"Tick\"}");
        Console.WriteLine($"one tick later:      last = {LastPrice(raw)}");
    }
}

File: examples/go/time_machine.go
// The time-machine scenario: rewind a recorded feed and watch state re-fold.
//
// Seek throws the folded state away and rebuilds it from the recording, so a
// rewind is deterministic rather than approximate -- which is what makes a
// recording more than a slow synthetic feed. Nothing here is Go-specific: it is
// four JSON commands, and every binding drives the same four.
//
//	cd examples/go && go run . time-machine
package main

import (
	"encoding/json"
	"fmt"
	"strings"

	wickra "github.com/wickra-lib/wickra-terminal-go"
)

const trades = 6

// The recorded feed, as the JSON array a Replay source takes.
func replayConfig() string {
	events := make([]string, 0, trades)
	for i := 0; i < trades; i++ {
		events = append(events, fmt.Sprintf(
			`{"type":"trade","symbol":{"base":"BTC","quote":"USDT"},`+
				`"price":"%d","quantity":"1","aggressor":"Buy","timestamp":%d}`,
			100+i, i+1))
	}
	feed, err := json.Marshal("[" + strings.Join(events, ",") + "]")
	if err != nil {
		panic(err)
	}
	return `{"sources":[{"Replay":{"dataset":` + string(feed) + `}}],` +
		`"layout":{"panels":[{"kind":"Chart","rect":{"x":0,"y":0,"w":100,"h":100}}]}}`
}

// The chart panel's last price, out of a frame.
func lastPrice(frame string) float64 {
	var decoded struct {
		Panels []struct {
			Panel string  `json:"panel"`
			Last  float64 `json:"last"`
		} `json:"panels"`
	}
	if err := json.Unmarshal([]byte(frame), &decoded); err != nil {
		panic(err)
	}
	for _, panel := range decoded.Panels {
		if panel.Panel == "chart" {
			return panel.Last
		}
	}
	return 0
}

func timeMachine() {
	term, err := wickra.New(replayConfig())
	if err != nil {
		panic(err)
	}
	defer term.Close()

	must := func(command string) string {
		out, err := term.Command(command)
		if err != nil {
			panic(err)
		}
		return out
	}

	must(`{"type":"Subscribe","source":0,"symbol":"BTC/USDT"}`)
	var raw string
	for i := 0; i < trades; i++ {
		raw = must(`{"type":"Tick"}`)
	}
	fmt.Printf("played to the end:   last = %v\n", lastPrice(raw))

	fmt.Println("position:           ", must(`{"type":"ReplayPosition","source":0}`))

	// Rewind to just after the second trade. The state is rebuilt from the
	// recording rather than restored from a snapshot, which is why a rewind
	// lands on exactly the frame the forward pass had at that point.
	raw = must(`{"type":"Seek","source":0,"index":2}`)
	fmt.Printf("rewound to index 2:  last = %v\n", lastPrice(raw))

	// And forward again from there, over the same events.
	raw = must(`{"type":"Tick"}`)
	fmt.Printf("one tick later:      last = %v\n", lastPrice(raw))
}
